### PR TITLE
Add mimetype config

### LINKED ISSUE #87: Přidat možnost vybrat databázi/knihovnu na mimetypes
Funkce `freeze` by měla brát nový argument, `guess_type`, což bude funkce, která se chová jako `mimetypes.guess_type` a použije se místo ní.

### PR DESCRIPTION
**Pull request by nappex**
fix #87

### CODE REVIEW

File: tests/test_check_mimetype.py [hunk 1]
```diff
@@ -2,157 +2,191 @@
...
+        [
+            ('coNtENT-TypE', 'IMAgE/JpEG'),
+            ('Content-Length', '654'),
+        ],
+    ),
 
-def test_case_insensitive_headers():
-    check_mimetype(
+    # there is diff between jpg and jpeg, our default frecognizer (guess_type)
+    # return jpg as jpeg.
```
Comment (encukou) [inline tests/test_check_mimetype.py:60]:
```suggestion
    # The content-type for JPEG images is standardized: `image/jpeg` (case insensitive).
    # The file extension is not standardized; the default recognizer guesses that
    # both `.jpg` or `.jpeg` mean JPEG images.
```

File: freezeyt/freezer.py [hunk 2]
```diff
@@ -59,11 +59,14 @@ async def freeze_async(app, config):
 }
 
 
-def check_mimetype(url_path, headers, default='application/octet-stream'):
+def check_mimetype(
+    url_path, headers,
+    frecognizer=guess_type, default='application/octet-stream'
```
Comment (encukou) [inline freezeyt/freezer.py:64]:
```suggestion
    default='application/octet-stream', *, frecognizer=guess_type,
```

File: freezeyt/freezer.py [hunk 3]
```diff
@@ -59,11 +59,14 @@ async def freeze_async(app, config):
...
+):
     if url_path.endswith('/'):
         # Directories get saved as index.html
         url_path = 'index.html'
-    file_type, file_encoding = guess_type(url_path)
+    file_type, file_encoding = frecognizer(url_path)
     if not file_type:
         file_type = default
```
Comment (encukou) [inline freezeyt/freezer.py:71]: When `None` is returned for `file_type`, the default is used. This should be in the documentation and tests.

File: README.md [hunk 4]
```diff
@@ -282,6 +284,31 @@ use:
 default_mimetype=text/plain
 ```
 
+#### File type recognition
```
Comment (encukou) [inline README.md:287]:
```suggestion
#### Guessing file types from extensions
```
+ use `guess_type` everywhere

File: README.md [hunk 5]
```diff
@@ -282,6 +284,31 @@ use:
...
+
+Value is recognizer, which can be defined as:
+* strings in the form `"module:function"`, which name the finder
+  function to call,
+* Python functions (if configuring `freezeyt` from Python, e.g. as a dict,
+  rather than YAML).
+
+
+An URL finder gets arguments:
+* filepath as `PurePath`
```
Comment (encukou) [inline README.md:305]: A string (the path's `name`) would be better for a public API.

Comment (nappex) [reply README.md]: Totally true I've updated

File: freezeyt/freezer.py (hunk 3 above)
Comment (nappex) [reply freezeyt/freezer.py:71]: - [ ] documentation
- [ ] tests

File: freezeyt/freezer.py [hunk 6]
```diff
@@ -394,8 +397,13 @@ def start_response(
         status_action = status_handler(hooks.TaskInfo(task, self))
 
         if status_action == 'save':
+            frecognizer = self.config.get('filetype_recognizer', guess_type)
+            if isinstance(frecognizer, str):
+                frecognizer = import_variable_from_module(frecognizer)
```
[the remark this answers is not in the document]
Comment (nappex) [reply freezeyt/freezer.py:402]: for sure done it

File: README.md [hunk 7]
```diff
@@ -282,6 +284,38 @@ use:
...
+
+`get_mimetype` can be defined as:
+* strings in the form `"module:function"`, which name the function to call,
+* Python functions (if configuring `freezeyt` from Python, e.g. as a `dict`,
+  rather than YAML).
+
+The `get_mimetype`:
+*gets one argument the `filepath` as `string`
+
+*returns file type as `string` (e.g. `"text/plain"`).
```
Comment (nappex) [inline README.md:314]: We have conflict with what should return our `get_mimetype` and what is returning func guess_mimetype.

File: freezeyt/freezer.py [hunk 8]
```diff
@@ -59,11 +59,14 @@ async def freeze_async(app, config):
...
-def check_mimetype(url_path, headers, default='application/octet-stream'):
+def check_mimetype(
+    url_path, headers,
+    default='application/octet-stream', *, get_mimetype=guess_type,
+):
     if url_path.endswith('/'):
         # Directories get saved as index.html
         url_path = 'index.html'
-    file_type, file_encoding = guess_type(url_path)
+    file_type, _ = get_mimetype(url_path)
```
Comment (nappex) [inline freezeyt/freezer.py:69]: How to handle our expectations about returning just a string but `guess_type` returning a `Tuple` of `strings`

File: README.md [hunk 9]
```diff
@@ -282,6 +284,38 @@ use:
...
+
+Freezeyt will register your own function, if you specify it in configuration
+YAML file as:
+
+```yaml
+get_mimetype=module:your_function
+```
+
+If the `get_mimetype` is not defined in configuration file,
+then the `freezeyt` uses the python function `mimetypes.guess_type`
```
Comment (encukou) [inline README.md:304]:
```suggestion
then `freezeyt` calls the python function `mimetypes.guess_type`
and uses the mimetype (the first element) it returns.
```

File: README.md [hunk 10]
```diff
@@ -282,6 +284,38 @@ use:
...
+* strings in the form `"module:function"`, which name the function to call,
+* Python functions (if configuring `freezeyt` from Python, e.g. as a `dict`,
+  rather than YAML).
+
+The `get_mimetype`:
+*gets one argument the `filepath` as `string`
+
+*returns file type as `string` (e.g. `"text/plain"`).
+
+If `mimetype` can't be determine, `get_mimetypes` returns `None`.
```
Comment (encukou) [inline README.md:316]:
```suggestion
If `get_mimetypes` returns `None`, `freezeyt` will use the configured `default_mimetype` (see *Default MIME type* above).
```

**Comment by nappex** [comment]
- [ ] make wrapper around get_mimetype
- [ ] try to refactor the loading keys from configuration

**Review by encukou — APPROVED** [review]

**Pull request merged**